### PR TITLE
core/main: fix setting of arguments for shutdown

### PR DESCRIPTION
**Pull request by keszybz**
Fixup for d2ebd50d7f9740dcf30e84efc75610af173967d2 and 6920049fad4fa39db5fec712f82f7f75b98fd4b9:
- add a comment that the last arg must be NULL and adjust the assert.
- move initialization around so that fields are declared, initialized, and consumed in the same order.
- move declaration of pos adjacent do declaration of command_line. This makes it easy to see that it was not initialized correctly.
- initialize buffers before writing the pointer into the args array. This makes no difference for the compiler, but it just feels "wrong" to do it in opposite order.

Because pos was off, we would ignore args after the timeout, and also overwrite the buffer if enough args were used.

I think this is case shows clearly that declaring all variables at the top of the function, with some initialized and other not, is very error-prone. The compiler has no issue with declaring variables whereever, and we should take advantage of this to make it keep declaration, initialization, and use close. (Within reason of course.)

### CODE REVIEW

**Review by yuwata — APPROVED** [review]
Ouch. LGTM. Thank you

File: src/core/main.c [hunk 1]
```diff
@@ -1586,6 +1589,10 @@ static int become_shutdown(int objective, int retval) {
         r = watchdog_setup(watchdog_timer);
         watchdog_close(r < 0);
 
+        /* The environment block: */
+
+        env_block = strv_copy(environ);
```
Comment (poettering) [inline src/core/main.c:1594]: hmm, no oom check?

File: src/core/main.c [hunk 2]
```diff
@@ -1567,11 +1568,13 @@ static int become_shutdown(int objective, int retval) {
                 command_line[pos++] = "--log-time";
 
         if (objective == MANAGER_EXIT) {
-                command_line[pos++] = exit_code;
                 xsprintf(exit_code, "--exit-code=%d", retval);
+                command_line[pos++] = exit_code;
         }
 
-        assert(pos < ELEMENTSOF(command_line));
+        assert(pos < ELEMENTSOF(command_line) - 1);
```
Comment (poettering) [inline src/core/main.c:1575]: hmm? why the -1?

File: src/core/main.c (hunk 1 above)
Comment (poettering) [reply src/core/main.c:1594]: hmm, we do no oom checks for the whole block. i guess that might be ok then, we can't do much anyway here

File: src/core/main.c (hunk 2 above)
Comment (poettering) [reply src/core/main.c:1575]: got it

**Comment by vcaputo** [comment]
Huh, my heart sank a bit when I saw these notifications in my inbox this morning, but it doesn't look like my compile-tested-only cleanup of this stuff was actually responsible?  I'll take it.

**Comment by keszybz** [comment]
Pffff. It turns out that I'm the idiot here. The only think I can say is that I managed to successfully fool all reviewers ;)

The initial off-by-one was introduced in c5673ed0de3bec38f68d8113d253842b47766e27. The assert I "fixed" was intentionally introduced in 26abdc73a212b90f7c4b71808a1028d2e87ab09f as a bugfix. And looking at the description of that commit, and recounting everything on my fingers at least a dozen times, I think my "fix" needs to be reverted.

**Pull request merged**